samples/nrf9160/nrf_cloud_mqtt_multi_service/src/application.c

```diff
@@ -176,8 +187,103 @@ static void on_location_update(const struct location_event_data * const location
 	}
 }
 
+/**
+ * @brief Receives general device messages from nRF Cloud, checks if they are AT command requests,
+ * and performs them if so, transmitting the modem response back to nRF Cloud.
+ *
+ * Try sending {"appId":"MODEM", "messageType":"CMD", "data":"AT+CGMR"}
+ * in the nRF Cloud Portal Terminal card.
+ *
+ * @param msg - The device message to check.
+ */
+static void handle_at_cmd_requests(const char *const msg)
+{
+	/* Attempt to parse the message as if it is JSON */
+	struct cJSON *msg_obj = cJSON_Parse(msg);
+
+	if (!msg_obj) {
+		/* The message isn't JSON or otherwise couldn't be parsed. */
+		LOG_DBG("A general topic device message of length %d could not be parsed.",
+			msg ? strlen(msg) : 0);
+		return;
+	}
+
+	/* Check that we are actually dealing with an AT command request */
+	char *msg_appid =
+		cJSON_GetStringValue(cJSON_GetObjectItem(msg_obj, NRF_CLOUD_JSON_APPID_KEY));
+	char *msg_type =
+		cJSON_GetStringValue(cJSON_GetObjectItem(msg_obj, NRF_CLOUD_JSON_MSG_TYPE_KEY));
+	if (!msg_appid || !msg_type ||
+	    (strcmp(msg_appid, NRF_CLOUD_JSON_APPID_VAL_MODEM)  != 0) ||
+	    (strcmp(msg_type,  NRF_CLOUD_JSON_MSG_TYPE_VAL_CMD) != 0)) {
+		goto cleanup;
+	}
+
+	/* If it is, extract the command string */
+	char *cmd =
+		cJSON_GetStringValue(cJSON_GetObjectItem(msg_obj, NRF_CLOUD_JSON_DATA_KEY));
+
+	if (!cmd) {
+		/* Missing or invalid command value will be treated as a blank command */
+		cmd = "";
+	}
+
+	/* Execute the command and receive the result */
+	LOG_DBG("Modem AT command requested: %s", cmd);
+	memset(at_req_resp_buf, 0, sizeof(at_req_resp_buf));
+	int err = nrf_modem_at_cmd(at_req_resp_buf, sizeof(at_req_resp_buf), cmd);
+
+	/* Trim \r\n from modem response for better readability in the portal. */
+	at_req_resp_buf[MAX(0, strlen(at_req_resp_buf) - 2)] = '\0';
+
+	/* If an error occurred with the request, report it */
+	if (err < 0) {
+		/* Negative error codes indicate an error with the modem lib itself, so the
+		 * response buffer will be empty (or filled with junk). Thus, we can print the
+		 * error message directly into it.
+		 */
+		snprintf(at_req_resp_buf, sizeof(at_req_resp_buf),
+			 "Error while processing AT command request: %d", err);
+		LOG_ERR("%s", at_req_resp_buf);
+	}
+
+	/* Free the old container object and create a new one to contain our response */
+	cJSON_Delete(msg_obj);
+	msg_obj = create_timestamped_device_message(
+		NRF_CLOUD_JSON_APPID_VAL_MODEM,
+		NRF_CLOUD_JSON_MSG_TYPE_VAL_DATA
+	);
+
+	if (!msg_obj) {
+		return;
+	}
+
+	/* Populate response with command result */
+	if (!cJSON_AddStringToObject(msg_obj, NRF_CLOUD_JSON_DATA_KEY, at_req_resp_buf)) {
```

Comment:
if you send a `AT%CMNG=1` to the device, the response will have quotes... does this properly escape the quotes?

Reply:
I get a mysterious modem error that doesn't occur when running the command manually... Looking into why that might be